### ISSUE TITLE
Nullable analysis of non-returning local functions

### ISSUE THREAD

**Issue by RikkiGibson** [dotnet/roslyn#45814]
**Version Used**: 7b4bfaaee

**Steps to Reproduce**:

```cs
string? s = null;
local1();
s.ToString(); // should not warn

void local1() => throw null!;
```
**Expected Behavior**: No nullable warnings after call to non-returning local function

**Actual Behavior**: Nullable warning after call to non-returning local function

It feels like when we decide what to do in this scenario, we should also confirm whether `[DoesNotReturn]`/`[DoesNotReturnIf]` are supported on local functions. It feels like if we do the proper analysis the attributes should not be necessary on local functions.

Relates to https://github.com/dotnet/roslyn/issues/52827 (nullability attributes on lambdas should be enforced)

**Comment by RikkiGibson** [comment]
This is pretty much another form of #45770 (inferences made from analyzing local functions affecting state at the call sites)

**Comment by RikkiGibson** [comment]
I think the solution here might be to just ensure that the flow analysis attributes such as DoesNotReturn work on local functions as well as possible.

**Comment by jcouv** [comment]
Resolved in 17.0p5
